ImGUI DPI Scaling

michaeleggers opened the pull request:
Trying to find a way to get ImGUI to display crisp without being super tiny in High DPI displays.

**WIP!**

DanielGibson committed:
Update libs/ktx/ to latest git

commit dac48df00c3af3d37 "Fix python deprecation warning. (#1018)"
from 2025-04-17

this is from after the 4.4.0 release (my previous state was from shortly
before that release)

DanielGibson committed:
libktx: use zstddeclib.c instead of full zstd.c

enough for our purposes and reduces the build time of a full texview
rebuild on my system from 9.3s to 3.6s (!!!) and reduces the size of
the texview binary (on Linux amd64, MinSizeRel build) from 2.8MB to
2.4MB. Apparently just recompiling zstd.c took more than 8 seconds
(zstddeclib.c: 1.8s)

DanielGibson committed:
Render Quads with (mostly) modern GL instead of immediate mode

.. and render all at once and pass the desired mip level to the shader
instead of setting it with glTexParameteri(trgt, GL_TEXTURE_*_LEVEL, ..)

Now I only have to replace glViewport(), glOrtho() etc
and (in the shader) gl_ModelViewProjectionMatrix, then I should be able
to use a OpenGL 3.2 Core context (instead of Compatibility), which
should help with running this on macOS.

DanielGibson committed:
Set transform matrix as uniform, use OpenGL 3.2 Core Profile

and tiny change to CMakeLists.txt that should make it build on macOS

michaeleggers committed:
fix: Bump glsl version string to match version 3.2

michaeleggers committed:
fix: Bump glsl version string to match version 3.2

michaeleggers committed:
Try to get better scaling for imgui on MacOS

michaeleggers committed:
Merge branch 'dev' of github.com:DanielGibson/texview into dev

DanielGibson committed:
Use vector font in ImGui, scale it by reloading with different size

should look better than upscaling the bitmap font by setting
ImGui::GetIO().FontGlobalScale

also added a call to glfwPollEvents() before getting the window scale,
because it's supposed to work around a bug on macOS

DanielGibson committed:
Hopefully fix rendering on High-DPI X11

Turns out that glfwGetWindowContentScale() does *not* always return
glfwGetFramebufferSize() / glfwGetWindowSize() - on High-DPI X11 (at
least KDE), framebuffer and window size are (most probably) identical,
but contentscale still returns whatever you configured, so you can scale
your UI accordingly.
TBH I wish it was like this on all platforms...

DanielGibson commented:
Thanks a lot!

I added some changes to the dev branch that should help (incl. using a vector font that should look better when scaled up than the default bitmap font) :-)

It would be interesting to see what effect `glfwWindowHint(GLFW_COCOA_RETINA_FRAMEBUFFER, GLFW_TRUE)` (or with `GLFW_FALSE`) have on the reported framebuffer and window sizes

michaeleggers committed:
Merge branch 'dev' of github.com:DanielGibson/texview into dev

michaeleggers commented:
> It would be interesting to see what effect `glfwWindowHint(GLFW_COCOA_RETINA_FRAMEBUFFER, GLFW_TRUE)` (or with `GLFW_FALSE`) have on the reported framebuffer and window sizes

I tried `glfwWindowHint(GLFW_COCOA_RETINA_FRAMEBUFFER, GLFW_TRUE)` and `GLFW_FALSE`.
`TRUE` always reports a framebuffer twice the size of window size. 
`FALSE` always reports the exact framebuffer and window size that was given to GLFW in `glfwCreateWindow`, so the scaling is always 1.0.

The latter is expected but a scaling of 2.0 makes no sense to me. I will check out sokol's solution.

EDIT: The vector font helps **a lot**:

<img width="1276" alt="image" src="https://github.com/user-attachments/assets/8d293558-44a0-4329-aa72-6cc2574b6bf6" />


I also found out that if I set my screen resolution to something non-scaled then the ratio is always 1.0 as well (even if  `GLFW_COCOA_RETINA_FRAMEBUFFER` is enabled). This "non-scaled" resolution isn't exposed as such in MacOS screen settings. One has to click "Show all resolutions":

<img width="490" alt="image" src="https://github.com/user-attachments/assets/21b383d5-9564-4f0f-9202-f8409991bc79" />

DanielGibson commented:
> FALSE always reports the exact framebuffer and window size that was given to GLFW in glfwCreateWindow, so the scaling is always 1.0.

But is one pixel in the window/framebuffer one pixel on the screen then, or is it scaled up by macOS?

> This "non-scaled" resolution isn't exposed as such in MacOS screen settings

Non-scaled or scaled?
Is the default resolution (that's shown **without** "show all resolutions") the native physical resolution of the display or something that would get scaled up?

In the end, IMHO the most important thing is that the framebuffer gets physical pixels, so textures are displayed as sharp as possible.
If the suggested scaling scaling of 2.0 is too big, we could just add something like 
```
#if __APPLE__
if(imguiScale >= 2.0)
    imguiScale *= 0.75;
#endif
```
after glfwGetWindowContentScale().
Or maybe `imguiScale -= 0.5;` or even `imguiScale = 1.5;` (or whatever other factors/values look right), I don't know if it's always 2.0 or what kind of values you'd get on other "retina" displays.
Or is there maybe a way to get the displays DPI? If that is actually correct, we could use that to calculate a more appropriate scaling factor

DanielGibson commented:
Actually, I just realized that on your screenshot the sidebar isn't scaled by factor 2, but factor 4 - even though ImGui Scale is clearly set to 2.
It should be about 280 pixels wide when there's no scaling.
On your screenshot it's 1006 pixels or so - which is almost 4 times as big (I assume that it's a bit less because currently only the font is scaled, but border sizes etc are not).

What does it look like if you set the scale factor to 1? Is it too small then or just right?

It appears like ImGui already does appropriate scaling itself, somehow, based on `glfwGetFramebufferSize()` and `glfwGetWindowSize()`. That works fine on Wayland and probably macOS, but not on X11 (at least with KDE), because there both have the same size even though it should be scaled.

This behavior wasn't immediately obvious to me because I configured scaling of KDE/Wayland on my Laptop to only 125%, which probably isn't really enough, esp. for ImGui where the text is relatively small (I think with 150% many programs looked too big/used too much space).

Anyway, I guess I should only apply the scaling from glfwGetWindowContentScale() to ImGui if it has a different value than framebuffer_size / window_size

DanielGibson commented:
Hmm just so I don't forget (I should probably create an ImGui bug about it): This ImGui behavior seems kinda unfortunate. 

Because it implies that the font gets scaled by ImGui. ~~(**Update: not as relevant** One of the first thing ImGui does in AddFont() is clamping the given font size to a full integer (which is why I round it before passing it to ImGui) - **before** it applies any scaling)~~

In fact it looks like the scaling - it's stored in `ImGuiIO::DisplayFramebufferScale` - is only used when rendering, to calculate the framebuffer size. So ImGui does all calculations etc in window coordinates, not in real pixels.

The font gets rasterized into a texture when it's loaded, and as it gets loaded in window coordinate size instead of FB pixel size, it'll get scaled when rendering and not look as great as it could when being rendered at the same size, with one pixel in the font texture being one pixel on the screen.
OTOH, as all sizes from ImGui get converted as well, things likely start at fractions of a pixel anyway and text might get blurry just from that even if the font pixels had the same size as display pixels..

DanielGibson committed:
Only automatically scale ImGui size if it doesn't already do that itself

ImGui scales everything by framebuffer_size / window_size
(window size is in logical points that are not the same as pixels on
 High-DPI screens, at least on some platforms).
so if framebuffersize / window_size == content_scale
(from glfwGetWindowContentScale) then we shouldn't scale the font
any further.
if however the content_scale is bigger than that ratio, for example on
KDE/X11 with window-scaling enabled, the font is scaled up accordingly
after all.

DanielGibson commented:
> Anyway, I guess I should only apply the scaling from glfwGetWindowContentScale() to ImGui if it has a different value than framebuffer_size / window_size

I just pushed a commit that should do that.

DanielGibson committed:
Some improvements to ImGui scaling

- use ImGui::GetIO().DisplayFramebufferScale instead of calculating
  imguiCoordScale myself
    => myGLFWwindowsizefun() not needed anymore
- Also scale Style, not just font (if ImGui doesn't scale itself)
    - For this: Put setting style into a function so the default style
      (including its sizes) can easily be restored

DanielGibson committed:
Make sidebar a bit neater

- Texture Info (format, size, ...) is hidden behind a TreeNode now
  so it only uses space in the sidebar if the user actually wants to
  see that information
- Renamed "Array Index" to "Layer" so that slider+label doesn't make the
  sidebar wider than it already is
- comment out that debug info showing GLFW sizes and scales

michaeleggers committed:
Merge remote-tracking branch 'upstream/dev' into dev

michaeleggers committed:
Merge branch 'dev' of github.com:DanielGibson/texview into dev

michaeleggers committed:
Merge branch 'dev' of github.com:michaeleggers/texview into dev

michaeleggers committed:
Use native OSX API to read out display resolution in pixels

DanielGibson committed:
Save imgui.ini to settings directory in users home-dir (Windows TODO)

for Windows this is just stubs for now, but it's implemented for
macOS and UNIX-likes

Thanks to Tom Kidd of MacSourcePorts.com for msputils.m which
implements functions that help doing this on macOS!

michaeleggers committed:
Merge remote-tracking branch 'upstream/dev' into dev

DanielGibson commented:
I guess this can be closed now - thanks a lot for helping to improve this! :)
